File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        rows = len(maze)
        cols = len(maze[0])
        visitied = [[False] * cols for r in range(rows)]
        queue = []
        
        dx = [0, 1, 0, -1]
        dy = [1, 0, -1, 0]
        
        queue.append([entrance[0], entrance[1], 0])
        
        while queue:
            position = queue.pop(0)
            x, y, dist = position[0], position[1], position[2]

            if x < 0 or y < 0 or x >= rows or y >= cols or visitied[x][y] == True or maze[x][y] == "+":
                continue
                
            if (x == 0 or y == 0 or x == rows - 1 or y == cols - 1) and dist > 0:
                return dist
            
            visitied[x][y] = True
            
            for i in range(len(dx)):
                queue.append([x + dx[i], y + dy[i], dist + 1])

        return -1
            
        
        

File: 1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.java
class Solution {
    public int nearestExit(char[][] maze, int[] entrance) {
        /*
            Algorithms & DS: #BFS, #Queue
            Time: O(N*M), Extra Space: O(N*M)
         */

        int[] dr = { -1, 0, 1, 0 }, dc = { 0, 1, 0, -1 };
        int m = maze.length, n = maze[0].length;
        Queue<int[]> q = new LinkedList<>();
        int depth = 0;

        q.offer(entrance);
        maze[entrance[0]][entrance[1]] = '+';
        

        while (!q.isEmpty()){
            int size = q.size();

            for(int i = 0; i < size; i++){
                int[] curr = q.poll();

                for(int j = 0; j < dr.length; j++){
                    int[] next = { curr[0] - dr[j], curr[1] - dc[j] };

                    if(next[0] < 0 || next[1] < 0 || next[0] >= m || next[1] >= n){
                        if(depth > 0){
                            return depth;
                        }
                    }
                    else if(maze[next[0]][next[1]] == '.'){
                        q.offer(next);
                        maze[next[0]][next[1]] = '+';
                    }
                }
            }

            depth++;
        }
        
        return -1;
    }
}